### ISSUE TITLE
For DetailList, when row is selected, narrator is not reading corresponding column headers along with the description of the table cells data

### ISSUE THREAD

**Issue by chakradhar2** [OfficeDev/office-ui-fabric-react#7560]
**Issue :**  For DetailList, when row is selected, narrator is not reading corresponding column headers along with the description of the table cells data.
Suppose we have table like: 

|Name | Value
|-- | --
|Item1 | 1
|Item2 | 2

#### Actual behavior:
If first row is selected, Currently it reads like: Item1 , 1 selected

#### Expected behavior:

 It should read row some thing like : Item1, Column Header Name, 1 Column header Value Selected.

It should also mention link if present, while reading the row. Suppose Item1 has link
Item1 Link, Column Header Name, 1 Column header Value.
 

You can try on in Edge browser:   https://developer.microsoft.com/en-us/fabric#/components/detailslist/basic
Requested priority: Normal

**Comment by Vitalius1** [comment]
@KevinTCoughlin @natalieethell 
Through your accessibility pass with DetailsList, have you come across the issue described above? s this is even a valid issue? Can you enlighten us a little on what is expected to be announced from a reader in this case? Thanks!

@betrue-final-final is the author of the issue correct? When selecting a row the reader should announce each column? I just wonder what if there are 10 columns (or 20)...

**Comment by betrue-final-final** [comment]
No, it should not read all of the columns, that would make it very hard to get through a list. There should be a shorthand readout. In OneDrive, for example, we read the name, file type, date modified and modified by. For a generic list, we could default to just the first column value, but it would be good to enable people to set more like we do in OneDrive.

**Comment by natalieethell** [comment]
I believe this issue is more asking about column **header** readout than just cell value readout. Right now, when a row is focused, the screen reader reads out all cell **values** in that row. So, for this more advanced example, the following is read out

**accdb file icon, Loremipsum.accdb, 5/23/2017, Lorem Ipsum, 39 KB**

This is the icon name, file name, date modified, modified by, and file size of the item, respectively. It does not, however, say

**accdb file icon, File icon, Loremipsum.accdb, Name, 5/23/2017, Date Modified, Lorem Ipsum, Modified By, 39 KB, File Size of the item**

If you want just a single cell value to be read out, with its column header, you can use the arrow keys to step through each cell.

@betrue-final-final Does it still make sense that we would not want to read out every column **header** when a row is in focus?

**Comment by betrue-final-final** [comment]
If you read out every single cell in a row, then you need to read out the label or column header of each cell as well.

However, I don't think you should read out every cell in the row. You should just read out the first cell value and name, or in the case of a file list, the first few columns. It's only on cell navigation that the other cells should be read, and when a cell is read, it's header should also be read.

**Comment by Vitalius1** [comment]
@natalieethell @KevinTCoughlin 
If I understand correctly @betrue-final-final says that each user should decide which cell values of the focused row along with their corresponding column header should be read. Is this something that can be achieved? Could this be a feature we want to add to the accessibility capabilities of the DetailsList?

**Comment by natalieethell** [comment]
Yeah, sounds like Ben is saying we want to be able to specify how many cell + column header values should be read from the row when the row is selected. I'm not sure exactly how this would be implemented off the top of my head, though. It seems like when focus is placed on the row, the screen reader (both Narrator and NVDA have this behavior) read out all of the row's contents. 

I took a look at OneDrive and it looks like they place the information to be read out by the screen reader in the row's aria label, as you can see below:

![image](https://user-images.githubusercontent.com/14134000/51151200-c8e44780-181e-11e9-9546-9ee2c531c54f.png)

In case the aria label is hard to read:

**Test Folder 4, Folder, Modified August 21, 2017, edited by Natalie Ethell, 0 items, Private**

That's why it appears to be reading out the column headers. Additionally, the screen reader will stop after reading the aria label value and maybe the row number + total number of rows in the list. 

I don't really see another way around this other than to add a custom aria label. But I'm interested to hear @KevinTCoughlin's thoughts.

**Comment by Vitalius1** [comment]
I think having the aria-label on the row set in this way gives the total control to the dev to decide what he wants to be read by the screen reader. As @betrue-final-final that different scenarios may require different screen reader output. Maybe the list has 20 columns and DetailsList consumers only want to read out the most relevant ones for their app's users. Who decides which ones and how many?

**Comment by betrue-final-final** [comment]
We should have a smart default of 2-3 columns but let the author change if needed.

**Comment by natalieethell** [comment]
Okay, at the moment I'm not sure how that would be possible, but I can investigate more early next week.

**Comment by natalieethell** [comment]
I played around with this issue a little more and it looks like it may require a deeper investigation of the role hierarchy of DetailsList. I should be able to get back into this next week.

**Comment by natalieethell** [comment]
Sorry for the major delay here! I discussed this issue with the team, and given what Ben said above about being able to customize the number of column headers read out, providing your own aria label as I mentioned above seems like the best option. 

Here's [a CodePen](https://codepen.io/naethell/pen/LaOWzJ) demonstrating how to do that. When focusing on a details row in the CodePen, you should hear **Name Item [number] Value [number]** read out in both Narrator and NVDA.

All you need to do is define an aria label for each item and then pass in a callback function to `getRowAriaLabel` that returns that row's aria label.

```tsx
getRowAriaLabel={ (item: any) => item.ariaLabel }
```

I hope that helps, @chakradhar2!

**Comment by anubhavaggarwal22** [comment]
any updates @natalieethell  ?

**Comment by natalieethell** [comment]
@anubhavaggarwal22 have you tried my proposed solution above?

> Sorry for the major delay here! I discussed this issue with the team, and given what Ben said above about being able to customize the number of column headers read out, providing your own aria label as I mentioned above seems like the best option.
> 
> Here's [a CodePen](https://codepen.io/naethell/pen/LaOWzJ) demonstrating how to do that. When focusing on a details row in the CodePen, you should hear **Name Item [number] Value [number]** read out in both Narrator and NVDA.
> 
> All you need to do is define an aria label for each item and then pass in a callback function to `getRowAriaLabel` that returns that row's aria label.
> 
> ```ts
> getRowAriaLabel={ (item: any) => item.ariaLabel }
> ```
> 
> I hope that helps, @chakradhar2!

**Comment by paulgildea** [comment]
@xugao Let's use this as the main tracking issue.
